File: autonomousDriving/Assets/Scripts/LaneDetect_v2.cs
using System.Collections.Generic;
using UnityEngine;
using OpenCvSharp;
using UnityEngine.UI;
using System.Linq;
using System;
using System.Collections;

public class LaneDetect_v2 : MonoBehaviour
{
    [SerializeField]
    private RawImage rawImage;

    Camera cam;

    Mat Process(Mat image)
    {
        Mat output = new();
        int height = image.Height;
        int width = image.Width;

        image.CopyTo(output);

        Point[] region_of_interest_vertices =
            {
            new Point(0, height),
            new Point(width * 0.2, height * 0.65),
            new Point(width * 0.8, height * 0.65),
            new Point(width, height)
        };

        Mat[] mats = Bird_eye_view(image, width, height, region_of_interest_vertices);

        Mat bv_crop = Color_filter(mats[0]);    //8UC1
        Mat histogram = Lane_histogram(bv_crop);

        Mat black = new(mats[0].Size(), MatType.CV_8UC3, Scalar.Black);

        Lane_peak(histogram, out Point leftbases, out Point rightbases);

        List<List<Point>> drawinfo = Slide_window_search(bv_crop, leftbases, rightbases);

        #region 좌표 그리기
        /*int left_idx = 0;
        int right_idx = 0;

        foreach (Point pt in region_of_interest_vertices)
        {
            Cv2.Circle(image, pt, 5, Scalar.Red);
        }

        foreach (Point pt in drawinfo[0])
        {
            Cv2.Circle(mats[0], pt, 5, Scalar.Red);
            Cv2.PutText(mats[0], left_idx.ToString(), pt, HersheyFonts.HersheySimplex, 1, Scalar.Red);
            left_idx++;
        }

        foreach (Point pt in drawinfo[1])
        {
            Cv2.Circle(mats[0], pt, 5, Scalar.Green);
            Cv2.PutText(mats[0], right_idx.ToString(), pt, HersheyFonts.HersheySimplex, 1, Scalar.Green);
            right_idx++;
        }*/
        #endregion

        // 그리기
        List<Point> drawinfo_all = new(drawinfo[1]);
        drawinfo[0].Reverse();
        drawinfo_all.AddRange(drawinfo[0]);

        Cv2.FillConvexPoly(black, drawinfo_all, new Scalar(0, 255, 0));

        Cv2.Polylines(black, drawinfo, false, new Scalar(0, 255, 255), 5);

        Cv2.WarpPerspective(black, output, mats[1], new Size(width, height));

        //return bv_crop;
        return output;
    }

    Mat[] Bird_eye_view(Mat img_frame, int width, int height, Point[] region_of_interest_vertices)
    {
        Mat b_v = new();

        Point2f left_bottom = region_of_interest_vertices[0];
        Point2f left_top = region_of_interest_vertices[1];
        Point2f right_top = region_of_interest_vertices[2];
        Point2f right_bottom = region_of_interest_vertices[3];

        List<Point2f> pts1 = new();
        List<Point2f> pts2 = new();

        pts1.Add(left_top);
        pts1.Add(left_bottom);
        pts1.Add(right_top);
        pts1.Add(right_bottom);

        pts2.Add(new Point2f(0, 0));
        pts2.Add(new Point2f(0, height));
        pts2.Add(new Point2f(width, 0));
        pts2.Add(new Point2f(width, height));

        // pts1의 좌표에 표시. perspective 변환 후 이동 점 확인
        Mat M = Cv2.GetPerspectiveTransform(pts1, pts2);
        Mat ret_m = Cv2.GetPerspectiveTransform(pts2, pts1);
        Cv2.WarpPerspective(img_frame, b_v, M, new Size(width, height));

        Mat[] mats = new Mat[2];

        mats[0] = b_v; // 버드뷰
        mats[1] = ret_m;

        return mats;
    }

    Mat Color_filter(Mat image)
    {
        Mat output = new();
        Mat hsv = new();

        Mat white_mask = new();
        Mat yellow_mask = new();
        Mat gray_image = new();

        Mat white_img = new();
        Mat yellow_img = new();
        Mat res = new();

        Cv2.CvtColor(image, hsv, ColorConversionCodes.BGR2HSV);

        Cv2.InRange(hsv, new(0, 0, 220), new(255, 70, 255), white_mask);
        Cv2.InRange(hsv, new(20, 30, 100), new(40, 255, 255), yellow_mask);
        
        Cv2.BitwiseAnd(image, image, white_img, white_mask);
        Cv2.BitwiseAnd(image, image, yellow_img, yellow_mask);
        Cv2.BitwiseOr(white_img, yellow_img, res);

        Cv2.CvtColor(res, gray_image, ColorConversionCodes.BGR2HSV);
        Cv2.Canny(gray_image, output, 100, 120);

        return output;
    }

    Mat Lane_histogram(Mat image)
    {
        Mat output = new();
        Mat cropped = new(image, new OpenCvSharp.Rect(0, image.Rows / 2, image.Cols, image.Rows / 2));
        Cv2.Reduce(cropped / 255, output, ReduceDimension.Row, ReduceTypes.Sum, 4);

        return output;
    }

    void Lane_peak(Mat image, out Point left_max_loc, out Point right_max_loc)
    {
        int midpoint = image.Cols / 2;
        Mat left_half = image.ColRange(0, midpoint);
        Mat right_half = image.ColRange(midpoint, image.Cols);

        int count_letf = 0;
        int sum_left = 0;
        int mean_left;

        int count_right = 0;
        int sum_right = 0;
        int mean_right;

        for (int i = 0; i < midpoint; i++)
        {
            if (left_half.At<int>(0, i) != 0)
            {
                sum_left += i;
                count_letf++;
            }

            if (right_half.At<int>(0, i) != 0)
            {
                sum_right += i;
                count_right++;
            }
        }

        if(count_letf > 0)
        {
            mean_left = sum_left / count_letf;
            left_max_loc = new Point(mean_left, 0);
        }
        else
        {
            Cv2.MinMaxLoc(left_half, out _, out left_max_loc);
        }

        if(count_right > 0)
        {
            mean_right = sum_right / count_right;
            right_max_loc = new Point(mean_right + midpoint, 0);
        }
        else
        {
            Cv2.MinMaxLoc(right_half, out _, out right_max_loc);
        }
    }

    List<List<Point>> Slide_window_search(Mat binary_warped, Point left_current, Point right_current)
    {
        int nwindows = 12;
        int window_height = binary_warped.Height / nwindows;

        List<Point> nonzero = new();

        Cv2.FindNonZero(binary_warped, OutputArray.Create(nonzero));

        int margin = 60;

        List<Point> left_lane = new();
        List<Point> right_lane = new();
        List<Point> new_left_lane = new();
        List<Point> new_right_lane = new();

        new_left_lane.Add(new Point(left_current.X, binary_warped.Height));
        new_right_lane.Add(new Point(right_current.X, binary_warped.Height));

        for (int w = 0; w < nwindows; w++)
        {
            int win_y_low = binary_warped.Height - (w + 1) * window_height;  // window 윗부분
            int win_xleft_low = left_current.X - margin;  // 왼쪽 window 왼쪽 위
            int win_xleft_high = left_current.X + margin; // 왼쪽 window 오른쪽 아래
            int win_xright_low = right_current.X - margin;  // 오른쪽 window 왼쪽 위
            int win_xright_high = right_current.X + margin; // 오른쪽 window 오른쪽 아래

            Point good_left = new();
            Point good_right = new();

            foreach (Point pt in nonzero)
            {
                if ((pt.Y == win_y_low) & (pt.X >= win_xleft_low) & (pt.X < win_xleft_high))
                {
                    good_left = new Point(pt.X, pt.Y);
                    left_lane.Add(good_left);
                }
            }

            foreach (Point pt in nonzero)
            {
                if ((pt.Y == win_y_low) & (pt.X >= win_xright_low) & (pt.X < win_xright_high))
                {
                    good_right = new Point(pt.X, pt.Y);
                    right_lane.Add(good_right);
                }
            }

            Point new_left = new();
            Point new_right = new();

            if (left_lane.Count > 0)
            {
                RePos_center(left_lane, ref new_left, good_left.Y);
                new_left_lane.Add(new_left);
            }
            /*else
            {
                new_left_lane.Add(new Point(new_left_lane.Last().X, win_y_low));
            }*/

            if (right_lane.Count > 0)
            {
                RePos_center(right_lane, ref new_right, good_right.Y);
                new_right_lane.Add(new_right);
            }

            left_lane.Clear();
            right_lane.Clear();
        }

        return new List<List<Point>> { new_left_lane, new_right_lane };
    }

    void RePos_center(List<Point> Pos, ref Point current, int height)
    {
        int sum = 0;

        foreach (Point pt in Pos)
        {
            sum += pt.X;
        }

         int mean = sum / Pos.Count;
         current = new Point(mean, height);
    }

    void Start()
    {
        cam = GetComponent<Camera>();
    }

    void Update()
    {
        Texture2D img_copy = cam.targetTexture.toTexture2D();
        Mat img_frame = OpenCvSharp.Unity.TextureToMat(img_copy);

        Mat img_result = Process(img_frame);

        rawImage.texture = OpenCvSharp.Unity.MatToTexture(img_result);

        Resources.UnloadUnusedAssets();
    }
}

File: autonomousDriving/Assets/Scripts/Python/LaneDetect_V2.py
import cv2
import numpy as np


def slide_window_search(binary_warped, left_current, right_current):
    out_img = np.dstack((binary_warped, binary_warped, binary_warped))
    nwindows = 12
    window_height = np.int32(binary_warped.shape[0] / nwindows)
    nonzero = binary_warped.nonzero()  # 선이 있는 부분의 인덱스만 저장
    nonzero_y = np.array(nonzero[0])  # 선이 있는 부분 y의 인덱스 값
    nonzero_x = np.array(nonzero[1])  # 선이 있는 부분 x의 인덱스 값
    margin = 100
    minpix = 50
    left_lane = []
    right_lane = []
    color = [0, 255, 0]
    thickness = 2

    for w in range(nwindows):
        win_y_low = binary_warped.shape[0] - (w + 1) * window_height  # window 윗부분
        win_y_high = binary_warped.shape[0] - w * window_height  # window 아랫 부분
        win_xleft_low = left_current - margin  # 왼쪽 window 왼쪽 위
        win_xleft_high = left_current + margin  # 왼쪽 window 오른쪽 아래
        win_xright_low = right_current - margin  # 오른쪽 window 왼쪽 위
        win_xright_high = right_current + margin  # 오른쪽 window 오른쪽 아래

        cv2.rectangle(out_img, (win_xleft_low, win_y_low), (win_xleft_high, win_y_high), color, thickness)
        cv2.rectangle(out_img, (win_xright_low, win_y_low), (win_xright_high, win_y_high), color, thickness)
        good_left = ((nonzero_y >= win_y_low) & (nonzero_y < win_y_high) & (nonzero_x >= win_xleft_low) & (nonzero_x < win_xleft_high)).nonzero()[0]
        good_right = ((nonzero_y >= win_y_low) & (nonzero_y < win_y_high) & (nonzero_x >= win_xright_low) & (nonzero_x < win_xright_high)).nonzero()[0]
        left_lane.append(good_left)
        right_lane.append(good_right)

        if len(good_left) > minpix:
            left_current = np.int32(np.mean(nonzero_x[good_left]))
        if len(good_right) > minpix:
            right_current = np.int32(np.mean(nonzero_x[good_right]))

    left_lane = np.concatenate(left_lane)  # np.concatenate() -> array를 1차원으로 합침
    right_lane = np.concatenate(right_lane)

    leftx = nonzero_x[left_lane]
    lefty = nonzero_y[left_lane]
    rightx = nonzero_x[right_lane]
    righty = nonzero_y[right_lane]

    left_fit = np.polyfit(lefty, leftx, 2)
    right_fit = np.polyfit(righty, rightx, 2)

    ploty = np.linspace(0, binary_warped.shape[0] - 1, binary_warped.shape[0])
    left_fitx = left_fit[0] * ploty ** 2 + left_fit[1] * ploty + left_fit[2]
    right_fitx = right_fit[0] * ploty ** 2 + right_fit[1] * ploty + right_fit[2]

    ltx = np.trunc(left_fitx)  # np.trunc() -> 소수점 부분을 버림
    rtx = np.trunc(right_fitx)

    out_img[nonzero_y[left_lane], nonzero_x[left_lane]] = [255, 0, 0]
    out_img[nonzero_y[right_lane], nonzero_x[right_lane]] = [0, 0, 255]

    ret = {'left_fitx': ltx, 'right_fitx': rtx, 'ploty': ploty}

    return ret


def plothistogram(image):
    histogram = np.sum(image[image.shape[0]//2:, :], axis=0)
    midpoint = np.int32(histogram.shape[0]/2)
    leftbase = np.argmax(histogram[:midpoint])
    rightbase = np.argmax(histogram[midpoint:]) + midpoint

    return leftbase, rightbase


def region_of_interest(img, vertices):
    mask = np.zeros_like(img)

    match_mask_color = 255
    cv2.fillPoly(mask, vertices, match_mask_color)
    masked_image = cv2.bitwise_and(img, mask)
    return masked_image


def bird_eye_view(frame, width, height, region_of_interest_vertices):

    left_bottom = region_of_interest_vertices[0]
    left_top = region_of_interest_vertices[1]
    right_top = region_of_interest_vertices[2]
    right_bottom = region_of_interest_vertices[3]

    pts1 = np.float32([[left_top, left_bottom, right_top, right_bottom]])
    # 좌표의 이동점
    pts2 = np.float32([[0, 0], [0, height], [width, 0], [width, height]])

    # pts1의 좌표에 표시. perspective 변환 후 이동 점 확인.
    M = cv2.getPerspectiveTransform(pts1, pts2)
    ret_m = cv2.getPerspectiveTransform(pts2, pts1)
    b_v = cv2.warpPerspective(frame, M, (width, height))

    return b_v, ret_m


def image_process(image):

    hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)

    upper_white = np.array([30, 255, 255])
    lower_white = np.array([0, 0, 200])

    upper_yellow = np.array([40, 255, 255])
    lower_yellow = np.array([20, 30, 100])

    mask_white = cv2.inRange(hsv, lower_white, upper_white)
    mask_yellow = cv2.inRange(hsv, lower_yellow, upper_yellow)

    res = cv2.bitwise_and(image, image, mask=mask_white)
    res2 = cv2.bitwise_and(image, image, mask=mask_yellow)
    res3 = cv2.bitwise_or(res, res2)

    gray_image = cv2.cvtColor(res3, cv2.COLOR_RGB2GRAY)
    canny_image = cv2.Canny(gray_image, 100, 120)

    return canny_image


def process(image):

    height = image.shape[0]
    width = image.shape[1]
    region_of_interest_vertices = [
        (0, height),
        (int(width*0.332),int(height*0.6)),
        (int(width * 0.65), int(height * 0.6)),
        (width, height)
    ]

    bv, ret_m = bird_eye_view(image, width, height, region_of_interest_vertices)
    bv_crop = image_process(bv)

    rightbase, leftbase = plothistogram(bv_crop)

    ret = slide_window_search(bv_crop, leftbase, rightbase)

    pts_mean, result = draw_lane_lines(image, bv, ret_m, ret)

    return result


def draw_lane_lines(original_image, warped_image, Minv, draw_info):
    left_fitx = draw_info['left_fitx']
    right_fitx = draw_info['right_fitx']
    ploty = draw_info['ploty']
    warp_zero = np.zeros_like(warped_image).astype(np.uint8)

    pts_left = np.array([np.transpose(np.vstack([left_fitx, ploty]))])
    pts_right = np.array([np.flipud(np.transpose(np.vstack([right_fitx, ploty])))])
    pts = np.hstack((pts_left, pts_right))

    mean_x = np.mean((left_fitx, right_fitx), axis=0)
    print(mean_x)
    pts_mean = np.array([np.flipud(np.transpose(np.vstack([mean_x, ploty])))])

    cv2.fillPoly(warp_zero, np.int32([pts]), color=(0, 255, 0))
    cv2.fillPoly(warp_zero, np.int32([pts_mean]), color=(0, 255, 0))
    cv2.polylines(warp_zero, np.int32([pts_right]), 0, (0, 255, 255), 20)
    cv2.polylines(warp_zero, np.int32([pts_left]), 0, (0, 255, 255), 20)

    newwarp = cv2.warpPerspective(warp_zero, Minv, (original_image.shape[1], original_image.shape[0]))
    # result = cv2.addWeighted(original_image, 1, newwarp, 1, 0)

    return pts_mean, newwarp


def detect_line(image_data):
        try:
            frame = process(image_data)
            cv2.imshow('LaneDetect', frame)
        except:
            pass